File: code_snippets/api-guide/model/java/prompter_workflow.java
package com.clarifai.example;

import com.clarifai.channel.ClarifaiChannel;
import com.clarifai.credentials.ClarifaiCallCredentials;
import com.clarifai.grpc.api.*;
import com.clarifai.grpc.api.status.StatusCode;

public class ClarifaiExample {

    //////////////////////////////////////////////////////////////////////////////////////////
    // In this section, we set the user authentication, app ID, and the details of the new
    // custom workflow. Change these strings to run your own example.
    /////////////////////////////////////////////////////////////////////////////////////////
    
    static final String USER_ID = "YOUR_USER_ID_HERE";
    //Your PAT (Personal Access Token) can be found in the portal under Authentication
    static final String PAT = "YOUR_PAT_HERE";
    static final String APP_ID = "YOUR_APP_ID_HERE";
    // Change these to create your own custom workflow
    static final String WORKFLOW_ID = "my-custom-prompter-workflow";

    static final String NODE_ID_1 = "prompter-model";
    static final String PROMPTER_MODEL_ID = "my-prompter-model";
    static final String PROMPTER_MODEL_USER_ID = "YOUR_USER_ID_HERE";
    static final String PROMPTER_MODEL_APP_ID = "my-custom-app";
    static final String PROMPTER_MODEL_VERSION_ID = "e851fb99a3b14df788ce11accee45c19";

    static final String NODE_ID_2 = "text-to-text";
    static final String LLM_MODEL_ID = "GPT-4";
    static final String LLM_MODEL_USER_ID = "openai";
    static final String LLM_MODEL_APP_ID = "chat-completion";
    static final String LLM_MODEL_VERSION = "5d7a50b44aec4a01a9c492c5a5fcf387";

    ///////////////////////////////////////////////////////////////////////////////////
    // YOU DO NOT NEED TO CHANGE ANYTHING BELOW THIS LINE TO RUN THIS EXAMPLE
    ///////////////////////////////////////////////////////////////////////////////////

    public static void main(String[] args) {

        V2Grpc.V2BlockingStub stub = V2Grpc.newBlockingStub(ClarifaiChannel.INSTANCE.getGrpcChannel())
                .withCallCredentials(new ClarifaiCallCredentials(PAT));

        MultiWorkflowResponse postWorkflowsResponse = stub.postWorkflows(
                PostWorkflowsRequest.newBuilder()
                        .setUserAppId(UserAppIDSet.newBuilder().setUserId(USER_ID).setAppId(APP_ID))
                        .addWorkflows(
                                Workflow.newBuilder()
                                        .setId(WORKFLOW_ID)
                                        .addNodes(
                                                WorkflowNode.newBuilder()
                                                        .setId(NODE_ID_1)
                                                        .setModel(
                                                                Model.newBuilder()
                                                                        .setId(PROMPTER_MODEL_ID)
                                                                        .setUserId(PROMPTER_MODEL_USER_ID)
                                                                        .setAppId(PROMPTER_MODEL_APP_ID)
                                                                        .setModelVersion(ModelVersion.newBuilder().setId(PROMPTER_MODEL_VERSION_ID))
                                                        )
                                        )
                                        .addNodes(
                                                WorkflowNode.newBuilder()
                                                        .setId(NODE_ID_2)
                                                        .setModel(
                                                                Model.newBuilder()
                                                                        .setId(LLM_MODEL_ID)
                                                                        .setUserId(LLM_MODEL_USER_ID)
                                                                        .setAppId(LLM_MODEL_APP_ID)
                                                                        .setModelVersion(ModelVersion.newBuilder().setId(LLM_MODEL_VERSION))
                                                        )
                                                        .addNodeInputs(NodeInput.newBuilder().setNodeId(NODE_ID_1))
                                        )
                        ).build()
        );

        if (postWorkflowsResponse.getStatus().getCode() != StatusCode.SUCCESS) {
            throw new RuntimeException("Post workflows failed, status: " + postWorkflowsResponse.getStatus());
        }

    }

}

File: code_snippets/api-guide/model/node/prompter_workflow.js
//index.js file

//////////////////////////////////////////////////////////////////////////////////////////
// In this section, we set the user authentication, app ID, and the details of the new
// custom workflow. Change these strings to run your own example.
/////////////////////////////////////////////////////////////////////////////////////////

const USER_ID = "YOUR_USER_ID_HERE";
// Your PAT (Personal Access Token) can be found in the portal under Authentification
const PAT = "YOUR_PAT_HERE";
const APP_ID = "YOUR_APP_ID_HERE";
// Change these to create your own custom workflow
const WORKFLOW_ID = "my-custom-prompter-workflow";

const NODE_ID_1 = "prompter-model";
const PROMPTER_MODEL_ID = "my-prompter-model";
const PROMPTER_MODEL_USER_ID = "YOUR_USER_ID_HERE";
const PROMPTER_MODEL_APP_ID = "my-custom-app";
const PROMPTER_MODEL_VERSION_ID = "e851fb99a3b14df788ce11accee45c19";

const NODE_ID_2 = "text-to-text";
const LLM_MODEL_ID = "GPT-4";
const LLM_MODEL_USER_ID = "openai";
const LLM_MODEL_APP_ID = "chat-completion";
const LLM_MODEL_VERSION = "5d7a50b44aec4a01a9c492c5a5fcf387";

/////////////////////////////////////////////////////////////////////////////
// YOU DO NOT NEED TO CHANGE ANYTHING BELOW THIS LINE TO RUN THIS EXAMPLE
/////////////////////////////////////////////////////////////////////////////

const { ClarifaiStub, grpc } = require("clarifai-nodejs-grpc");

const stub = ClarifaiStub.grpc();

// This will be used by every Clarifai endpoint call
const metadata = new grpc.Metadata();
metadata.set("authorization", "Key " + PAT);

stub.PostWorkflows(
    {
        user_app_id: {
            "user_id": USER_ID,
            "app_id": APP_ID
        },
        workflows: [
            {
                id: WORKFLOW_ID,
                nodes: [
                    {
                        id: NODE_ID_1,
                        model: {
                            id: PROMPTER_MODEL_ID,
                            user_id: PROMPTER_MODEL_USER_ID,
                            app_id: PROMPTER_MODEL_APP_ID,
                            model_version: {
                                id: PROMPTER_MODEL_VERSION_ID
                            }
                        }
                    },
                    {
                        id: NODE_ID_2,
                        model: {
                            id: LLM_MODEL_ID,
                            user_id: LLM_MODEL_USER_ID,
                            app_id: LLM_MODEL_APP_ID,
                            model_version: {
                                id: LLM_MODEL_VERSION
                            }
                        },
                        node_inputs: [
                            {
                                node_id: NODE_ID_1 
                            }
                        ]
                    }
                ]
            }
        ]
    },
    metadata,
    (err, response) => {
        if (err) {
            throw new Error(err);
        }

        if (response.status.code !== 10000) {
            throw new Error("Post workflows failed, status: " + response.status.description);
        }

        console.log(response);
    }
);